### ISSUE TITLE
JOSE support

### ISSUE THREAD

**Issue by whyoleg** [whyoleg/cryptography-kotlin#16]
Introduce standalone JOSE support module with JWT/JWK/etc.
Support for JWK encoding/decoding can also be used afterwards in providers other than WebCrypto.

Full list of RFCs:
- JWS: https://datatracker.ietf.org/doc/html/rfc7515
- JWE: https://datatracker.ietf.org/doc/html/rfc7516
- JWK: https://datatracker.ietf.org/doc/html/rfc7517
- JWA: https://datatracker.ietf.org/doc/html/rfc7518
- JWT: https://datatracker.ietf.org/doc/html/rfc7519
- JOSE examples: https://datatracker.ietf.org/doc/html/rfc7520

**Comment by malhobayyeb** [comment]
Up

**Comment by aschulz90** [comment]
I am currently in the progress of implementing a JWT library for Kotlin Multiplatform supporting all platforms this library also supports and have signing and verification done very easily so far. 
Next step is JWKS support, but is is currently blocked by missing support for decoding public keys using JWK format.

The current progress is avaiable here: https://github.com/Appstractive/jwt-kt
The JWK handling is implemented [here](https://github.com/Appstractive/jwt-kt/blob/master/jwt-jwks/src/commonMain/kotlin/com/appstractive/jwt/JSONWebKeyExt.kt#L25). 

Maybe the repository will be useful to you by using it as a playground to test the JWK implementation. 
There is a [test in the jwt-jwks module](https://github.com/Appstractive/jwt-kt/blob/master/jwt-jwks/src/commonTest/kotlin/jwt/JwksVerifierTests.kt), that should pass, once JWK decoding works for all platforms (pending interface changes).

**Comment by babisRoutis** [comment]
> Hey, @waltkb, thank you for the interest! That would be really nice! I've created a branch [jose](https://github.com/whyoleg/cryptography-kotlin/tree/jose) (from main), and we could start creating PR's there as a start. What I wanted from the initial version is to implement JWT and JWK support. I have no time to investigate the API which would fit Kotlin, so overall, it is up to you to start, and I will review PRs as we go. To begin, let us add `jose` module under `cryptography-serialization` folder. You can also take a look on setup of `asn1` module as a base for it — all targets + kotlinx-serialization-json. No other dependencies, at least for now. The main idea would be to be simple enough and use kotlinx-serialization-json to handle serialisation (obviously). Also, we will need `Base64` - I would propose to just use the one which is already available in kotlin-stdlib. It is experimental, but I believe it should be fine. If you have any other questions, feel free to ask here or anywhere else linked on GitHub!
> 
> When we have all the base blocks from the serialisation part, then we can start thinking about how to integrate it with signing part of `core` module to have the best experience.
> 
> Thank you once more, I'm looking forward to it!

Dear @whyoleg 

The branch that you mentioned (jose) is not present. Was this postponed for a future release?
You proposed to add a `jose` module under  `cryptography-serialization` folder. 
To my understanding similar modules like der / pem provide some functions to the rest of the modules. Thus, they don't depend on other (project) modules.

Can this be the case of jose?
For instance, I was thinking that to support JWK, the jose module should provide a (JSON Serializer) for ECKeys, thus it should depend on the core. An alternative could be to define some serializable classes (for instance ECPubKeyJWK) in jose a let somebody else to use them.

Can you please elaborate?

**Comment by whyoleg** [comment]
Hey @babisRoutis !

> The branch that you mentioned (jose) is not present.

Yeah, looks like I deleted it while preparing the release. Anyway, it was just the same as `main` at that point to initiate discussion/implementation, as I would think, that probably it will take not a single PR to complete. I could create it once more, if you do want to contribute :)

> Was this postponed for a future release?

It was planned for 0.5.0 (next "major") release for a long time, as I had other things to do.
I've just released 0.4.0 and now I need to switch to other projects for some time, so I'm not planning to work on it right.
But this is still on my radar as the first feature for the next major release.

> Can this be the case of Jose?

Yeah, sure, I think so! At least initially we should have all data models for serialization (similar to how ASN.1/DER). Only after that we could start thinking about integration.
The story with JWK is rather easy, as in this case `cryptography-serialization-jose` will be just an implementation detail of providers. JWK is already supported as a key format in `core` API.

The story with JWT is harder, as it will probably need more tight integration with `core` APIs like signatures. For this, I think, that may be we will be able to create an additional module, a.k.a. `cryptography-jwk` where there will be high-level API to work with JWT and cryptography which under the hood will use `cryptography-serialization-jose`.

Do you get the idea?
It's not final, but at least this is what I have in mind.
If you are willing to help with contribution on this or do have some specific use-cases apart from JWK support, I would like to hear them and connect here/slack/call to discuss if we can cooperate on this.


P.S. The library size, complexity and amount of moving parts is only increasing and so working on it alone becomes harder and harder so I would really like if the work could be divided :)

**Comment by whyoleg** [comment]
Hey @aschulz90 !
Thanks for your work and links! This is really useful.
BTW, may be you do want to somehow cooperate on JWK support in `cryptography-kotlin` based on the plan in the message above?

**Comment by whyoleg** [comment]
On additional note, last time I thought about how JWK is implemented currently in the library, I was not very sure if it's a good idea to have `JWK` embedded into `encode/decode` APIs.
The reason for its existence is because `WebCrypto` supports `JWK` out-of-the-box. Google's [tink](https://developers.google.com/tink) also supports JWK, so I thought it looks natural.
But now, when I was tried to add some of the algorithms, it filled a bit unnatural, as DER/PEM encoding in some cases needs less parameters, then needed for `JWK`.
Also, as `JWK` specify algorithm inside key, it would be nice to have an ability to somehow create signer/verifier based on just key without a requirement to specify expected algorithm (or digest) first.
I don't yet know if it possible to do this in a nice way. That's why having `cryptography-serialization-jose` independent from `core` would be a nice thing, as we could easier experiment with different ways of encoding/decoding of keys. May be in future, `JWK` will even be removed from `encode/decode` and used in another way.

I don't know yet, just leaving my thoughts here.
P.S. also, there is COSE, for which I don't have plans yet, but it would be nice to keep in mind on how it should be integrated on par with JOSE.

**Comment by babisRoutis** [comment]
> But now, when I was tried to add some of the algorithms, it filled a bit unnatural, as DER/PEM encoding in some cases needs less parameters, then needed for `JWK`. Also, as `JWK` specify algorithm inside key, it would be nice to have an ability to somehow create signer/verifier based on just key without a requirement to specify expected algorithm (or digest) first. I don't yet know if it possible to do this in a nice way. That's why having `cryptography-serialization-jose` independent from `core` would be a nice thing, as we could easier experiment with different ways of encoding/decoding of keys. May be in future, `JWK` will even be removed from `encode/decode` and used in another way.

This is the reason that triggered my question. 

To my understanding, having a no-deps JWK serialization module, means that `core` would need to adapt it, yet JOSE(and  COSE) come with their own quirks  that eventually will "pollute" `core`.

Perhaps, `jose` could be a top-level module, or even a new lib, that just consumes `cryptography-kotlin` (similar to @aschulz90 work).

**Comment by aschulz90** [comment]
> Thanks for your work and links! This is really useful.
BTW, may be you do want to somehow cooperate on JWK support in cryptography-kotlin based on the plan in the message above?

I have started to look into it some time ago, but I am not that deep into cryptography and was kinda ovewhelmed by the RFC and its terminology. From what I gathered, the model could look like this:

```kotlin
import dev.whyoleg.cryptography.CryptographyAlgorithmId
import dev.whyoleg.cryptography.algorithms.digest.Digest
import dev.whyoleg.cryptography.algorithms.digest.SHA256
import dev.whyoleg.cryptography.algorithms.digest.SHA384
import dev.whyoleg.cryptography.algorithms.digest.SHA512
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator


enum class Algorithm {
  HS256,
  HS384,
  HS512,
  RS256,
  RS384,
  RS512,
  PS256,
  PS384,
  PS512,
  EC256,
  EC384,
  EC512,
}

val Algorithm.digest: CryptographyAlgorithmId<Digest>
  get() =
      when (this) {
        Algorithm.EC256,
        Algorithm.PS256,
        Algorithm.RS256,
        Algorithm.HS256 -> SHA256

        Algorithm.EC384,
        Algorithm.RS384,
        Algorithm.PS384,
        Algorithm.HS384 -> SHA384

        Algorithm.EC512,
        Algorithm.HS512,
        Algorithm.RS512,
        Algorithm.PS512 -> SHA512
      }

enum class Curve {
  @SerialName("P-256") P256,
  @SerialName("P-384") P384,
  @SerialName("P-521") P521,
}

@OptIn(ExperimentalSerializationApi::class)
@Serializable
@JsonClassDiscriminator("kty")
sealed interface JSONWebKey {
  val alg: Algorithm?
  val kid: String?
}

@Serializable
@SerialName("RSA")
data class JSONWebKeyRSA(
    override val alg: Algorithm? = null,
    override val kid: String,
    val use: String,
    val n: String,
    val e: String,
    val d: String? = null,
    val p: String? = null,
    val q: String? = null,
    val dp: String? = null,
    val dq: String? = null,
    val qi: String? = null,
) : JSONWebKey

@Serializable
@SerialName("EC")
data class JSONWebKeyEC(
  override val alg: Algorithm? = null,
  override val kid: String,
  val crv: Curve,
  val x: String,
  val y: String,
  val d: String,
) : JSONWebKey

@Serializable
@SerialName("oct")
data class JSONWebKeyHMAC(
    override val alg: Algorithm? = null,
    override val kid: String,
    val k: String,
) : JSONWebKey

@Serializable
data class JSONWebKeySet(
    val keys: List<JSONWebKey>,
)
```

But I am not sure about the parameters, or if they even are all required to genrate the public key from.

> it would be nice to have an ability to somehow create signer/verifier based on just key without a requirement to specify expected algorithm (or digest) first.

This is how I currently generate the verifier based on the above model class: [JSONWebKeyExt.kt](https://github.com/Appstractive/jwt-kt/blob/master/jwt-jwks/src/commonMain/kotlin/com/appstractive/jwt/JSONWebKeyExt.kt#L22)

As for the actual generation of public keys: at least on JVM it seems to be pretty simple, e.g. like this: [jwks-rsa-java ](https://github.com/auth0/jwks-rsa-java/blob/master/src/main/java/com/auth0/jwk/Jwk.java#L176)

**Comment by babisRoutis** [comment]
> @whyoleg thanks for the summary!
> 
> > Probably in the end it could be just implementation detail, but also could be exposed to end users (TBD).
> 
> In a backend context, it would be nice to have the model classes exposed. Either for providing your own JWKS endpoint (if this library also supports public key -> JWKS conversion) or fetching the JWKS from another endpoint (but one could also simply pass the response body text/bytes without deserialization).

Totally agree. 

Not only for back-end, though. For instance, in the context of [EUDI Wallet](https://github.com/eu-digital-identity-wallet) there are several cases where working with JWKs is required on mobile side.

**Comment by whyoleg** [comment]
Yeah, I understand that. I was not clear on what I mean by `implementation detail`. Those models should be definitely exposed by `cryptography-serialization-jose`, but probably could be just an implementation details for points 2 and 3 - it depends.
So yes, those use-cases will be supported.

**Comment by babisRoutis** [comment]
Hi @whyoleg 

For what is worth 

I am building my own JOSE lib (early stage, not public yet). 
Integration with Cryptography Kotlin is very smooth, I think. (I miss X509Certificate support for those `x5c` 😄 )

```kotlin
context(_: Raise<ProvisionVerifyOperation.Error>, _: Json, _: CryptographyProvider)
public suspend fun JsonWebKey.operationVerify(algorithm: SignatureAlgorithm): Operation.Verify {
    val verifier = when (algorithm) {
        is SignatureAlgorithm.ECDSA if this is ECJsonWebKey -> signatureVerifierECDSA(algorithm)
        is SignatureAlgorithm.RSA if this is RSAJsonWebKey -> signatureVerifierRSA(algorithm)
        is SignatureAlgorithm.EdDSA if this is OctetKeyPairJsonWebKey -> signatureVerifierEdDSA()
        is SignatureAlgorithm.ECDSA -> raise(ProvisionVerifyOperation.Error.InvalidKeyType(algorithm, KeyType.EC, keyType))
        is SignatureAlgorithm.EdDSA -> raise(ProvisionVerifyOperation.Error.InvalidKeyType(algorithm, KeyType.OKP, keyType))
        is SignatureAlgorithm.RSA -> raise(ProvisionVerifyOperation.Error.InvalidKeyType(algorithm, KeyType.RSA, keyType))
        @Suppress("DEPRECATION") SignatureAlgorithm.None,
        is SignatureAlgorithm.HMAC -> raise(ProvisionVerifyOperation.Error.UnsupportedAlgorithm(algorithm))

    }
    return Operation.Verify { data, signature -> verifier.tryVerifySignature(data, signature) }
}

context(_: Raise<ProvisionVerifyOperation.Error>, json: Json, provider: CryptographyProvider)
private suspend fun ECJsonWebKey.signatureVerifierECDSA(algorithm: SignatureAlgorithm.ECDSA): SignatureVerifier {
    val cryptoCurve = algorithm.curve().toCryptoCurve()
    val digest = algorithm.toDigest()
    val jwkBytes = json.encodeToString(JsonWebKey.serializer(), this).encodeToByteArray()
    val pubKey = provider.get(ECDSA)
        .publicKeyDecoder(cryptoCurve)
        .decodeFromByteArray(EC.PublicKey.Format.JWK, jwkBytes)
    return pubKey.signatureVerifier(digest, ECDSA.SignatureFormat.RAW)
}

```

**Comment by whyoleg** [comment]
@babisRoutis, thanks for letting me know!

> I am building my own JOSE lib

I think that I've already seen several existing ones. So I'm just curious: is there a reason you create one more?
But, yeah, I think that integration like this is rather easy to do (and that's nice!), especially since [0.6.0](https://github.com/whyoleg/cryptography-kotlin/releases/tag/0.6.0) now supports JWK everywhere.

---

BTW, last year I was prototyping the JOSE (and a bit of COSE) support: https://github.com/whyoleg/cryptography-kotlin/commit/80dbde92b37c1df9547d1db4c248c25e06de93aa.
Overall, JOSE API is more or less good™, but I haven't had time since then to finalize prototyping COSE, so that the concepts and APIs are similar between them, as well as research or prototype SD-JWT, JWP, and some other recent additions to the cryptography serialization standards, so that the API is flexible enough and futureproof.

**Comment by babisRoutis** [comment]
Hi @whyoleg 

For SD-JWT (and SD-JWT-VC) I have implemented [this](https://github.com/eu-digital-identity-wallet/eudi-lib-jvm-sdjwt-kt) one
Project targets, currently, JVM yet 90% of the code is not jvm-specific.

The above declares SD-JWT as a parameterized-type like bellow
```kotlin
data class SdJwt<out JWT>(val jwt: JWT,    val disclosures: List<Disclosure>)
```
That's the driver behind my JOSE effort.: To replace the JWT parameter with an rfc compliant and kmp jwt.

Thank you for sharing the prototype. There are some very good ergononomic ideas.

**Comment by babisRoutis** [comment]
@whyoleg 

I took a better look at your `dev/jose` branch. I was not aware that you are actively working on JOSE.
Indeed,   perhaps, makes little sense for yet another JOSE 

If you want a helping hand on this, I have some questions

- Is `dev/jose` a  target for PRs related to jose or just a  prototype branch? 
- Would it be possible to assemble list of issues for this?

From the jose prototype I got the impression that operations like signing and verification are closely related to the pertinent keys (Jwk). 
Is in scope of the jose public api to cover cases where the library cannot have access to the key itself ?
For instance, keys managed by a secure enclave, or a remote  HSM etc.

**Comment by whyoleg** [comment]
@babisRoutis 

> I was not aware that you are actively working on JOSE.

Currently, I'm not. The last commit was a year ago, and after that, I haven't explored anything.

> Indeed, perhaps, makes little sense for yet another JOSE

Overall, I think it's fine to have several libraries, as different use cases might need different API designs and might not always need the full JOSE spec. But, for CK, the story is different. If JOSE is available out of the box, then it should be generic enough to be used in different use cases. And so, this means that API design is hard. :)
Implementing just JWT handling via explicit sign/verify APIs from CK should not be a big deal, and you already have everything in place in the repository you shared.

> Is dev/jose a target for PRs related to jose or just a prototype branch?

It's just a prototype, which was just stopped - TBH, I don't remember all of the details that need to be changed there. Maybe the main JOSE APIs are already there, and all we need is to add a minimal implementation.

> Would it be possible to assemble list of issues for this?

Let me think, what I can do, but all of those will be really about prototyping the API first, e.g.:
- design COSE similar to JOSE.
- design how SD-JWT will be integrated, considering how `JwtClaims` are implemented.
- prototype how the latest active [jose](https://datatracker.ietf.org/wg/jose/documents/) RFCs will look like, specifically "JSON Web Proof' and "JSON Proof Token and CBOR Proof Token". Those should not be integrated, but it should be possible to design a consistent API.
- check if we can also support https://datatracker.ietf.org/doc/html/rfc7797
- check how we can support compression in JWE, without implementing it in CK
- check if we can support not only `ByteArray` as a payload, but also a stream - could be useful in case of "[detached payload](https://datatracker.ietf.org/doc/html/rfc7797#section-5.1)" case

The implementation of all this is more or less simple, given that the spec is already in place. Creating nice Kotlin APIs is hard...

And in the end, it would be nice to validate all those APIs against a real-world use case. I was thinking of implementing the Ktor server/client plugin for JWT authorization at least. Maybe something related to VC.

> From the jose prototype I got the impression that operations like signing and verification are closely related to the pertinent keys (Jwk).

Not really. [They](https://github.com/whyoleg/cryptography-kotlin/blob/80dbde92b37c1df9547d1db4c248c25e06de93aa/cryptography-serialization/jose/src/commonMain/kotlin/JwsSigning.kt#L8-L9) just need headers. So if, based on the header, you want to use a non-exportable key, it's possible.

> Is in scope of the jose public api to cover cases where the library cannot have access to the key itself ? For instance, keys managed by a secure enclave, or a remote HSM etc.

Yes, it is. It's also within the scope of the CK to support such keys at some point for various cryptographic operations. :)
